statsd/metrics.go

```diff
@@ -123,59 +124,123 @@ func (s *setMetric) flushUnsafe() []metric {
 type bufferedMetric struct {
 	sync.Mutex
 
+	// Kept samples (after sampling)
 	data []float64
+	// Total stored samples (after sampling)
+	storedSamples int64
+	// Total number of observed samples (before sampling). This is used to keep
+	// the sampling rate correct.
+	totalSamples int64
+
 	name string
 	// Histograms and Distributions store tags as one string since we need
 	// to compute its size multiple time when serializing.
 	tags  string
 	mtype metricType
+
+	// maxSamples is the maximum number of samples we keep in memory
+	maxSamples int64
 }
 
 func (s *bufferedMetric) sample(v float64) {
 	s.Lock()
 	defer s.Unlock()
+	s.sampleUnsafe(v)
+}
+
+func (s *bufferedMetric) sampleUnsafe(v float64) {
 	s.data = append(s.data, v)
+	s.storedSamples++
+	atomic.AddInt64(&s.totalSamples, 1)
+}
+
+func (s *bufferedMetric) maybeKeepSample(v float64, r *rand.Rand, lock *sync.Mutex) {
+	s.Lock()
+	defer s.Unlock()
+	if s.maxSamples > 0 {
+		if s.storedSamples >= s.maxSamples {
+			// We reached the maximum number of samples we can keep in memory, so we randomly
+			// replace a sample.
+			r := r.Int63n(s.storedSamples)
+			if r < s.maxSamples {
```

Comment:
**question:** Which corner case is this guard handling?

Reply:
This makes more sense once we fix the call above with `r.Int63n(s.totalSamples)`